#!/DCEG/Resources/Tools/R/R-3.1.2/bin/Rscript
# args <- commandArgs(TRUE)

# R pcakge SKAT should be install before running burden test

VCF_DIR = '/DCEG/Projects/HPV_Consortium/HPV31_PAP_RD159_merged_single-uni_20160216/tvc_ann'
# VCF_DIR: vcf files' absolute directory

VCF_FILE_POST = '.ann.vcf'
# VCF_FILE_POST: vcf file name's post, for example, if vcf file name is "PAP144203.ann.vcf", then vcf.file.post = '.ann.vcf'; if vcf file name is "PAP144203.vcf", then vcf.file.post = '.vcf'
# vcf file name should be SAMPLE(">" lines from .fasta file) + vcf.file.post, for example, PAP144203.ann.vcf, where PAP144203 is from .fasta file
FASTA_FILE = '/DCEG/Projects/BB/KaiYu/HPV31_201702/HPV31_PAP_RD159_merged_ffpe-uni_20160217.fasta'

To_EXCLUDE_SAMPLES_FILE = NULL
# To_EXCLUDE_SAMPLES_FILE is NULL or a text file with one SAMPLE name each line, where SAMPLE should be available in .fasta file

GENE_INFO_FILE = '/DCEG/Projects/BB/KaiYu/HPV31_201702/HPV31.gene.start.end.txt'
# for example, HPV31:
# Gene	START	END
# URR	1	107
# E6	108	557
# E7	560	856
# E1	862	2751
# E2	2693	3811
# E4	3270	3578
# E5	3816	4070
# L2	4171	5571
# L1	5552	7066
# URR	7067	7912

EXTRACT_OUT_DIR = '/DCEG/Projects/BB/KaiYu/HPV31_201702/Burden'
# EXTRACT_OUT_DIR: output directory for the extracted info from vcf files
EXTRACT_OUT_RDATA_FILE = 'functional.class.info.genotype.Rdata'
# EXTRACT_OUT_RDATA_FILE: output rdata file (generated by calling extract.info()) which saved the extracted info from vcf files

PHENO_FILE = '/DCEG/Projects/BB/KaiYu/HPV31_201702/HPV31_pheno_small.txt'
# PHENO_FILE: pheno type file is tab-delimted text file
STATUS_NAME_IN_PHENO_FILE = 'binary_CIN3p_clean'
# STATUS_NAME_IN_PHENO_FILE: pheno type(used for burden test) column name
LINEAGES_NAME_IN_PHENO_FILE = 'main_lineage'
# LINEAGES_NAME_IN_PHENO_FILE: lineages column name
LINEAGES = c("AB","C")

MIN_MAF_THR = 0
# MIN_MAF_THR = minimum maf threshold
MAX_MAF_THR = 0.03
# MAX_MAF_THR = maximum maf threshold
PATTERNS = c("MISSENSE", "NONSENSE")


BURDEN_RESULTS_DIR = paste0(EXTRACT_OUT_DIR, '/Results')
# BURDEN_RESULTS_DIR: burden test results output directory

FINAL_HITS_PHENO_RDATA_FILE_POST = '_min.maf.0_max.maf.0.03_MISSENSE_NONSENSE_AB_C.Rdata'
# FINAL_HITS_PHENO_RDATA_FILE_POST: file post for each gene's output burden test Rdata file

extract.functional.class.genotype <- function(vcf.file, gene.info, out, vcf.file.post = '.ann.vcf'){
  # browser()
  n.line = as.numeric(system(paste('wc -l', vcf.file, "| awk '{print $1}'"), intern = T))
  if(n.line < 1)
    return(NULL)
  n.skip = as.numeric(system(paste('grep -n "#CHROM"', vcf.file, '| cut -d ":" -f 1'), intern = T))
  if(length(n.skip) < 1)
    return(NULL)
  if(n.skip >= n.line)
    return(NULL)
  
  genes = unique(gene.info[,1])
  vcf = read.delim(file = vcf.file, skip = n.skip - 1, header = T, sep = '\t', check.names = F, stringsAsFactors = F)
  
  row.idx = match(gsub(vcf.file.post, '', vcf.file), out[[1]][[1]][,1])
  n.row = nrow(out[[1]][[1]])
  
  for(i in 1:nrow(vcf)){
    cat(i, '\t')
    cur.info = strsplit(vcf[i, 8], ';')[[1]]
    if(length(cur.info) < 1)
      next
    
    cur.pos = vcf[i, 2]
    cur.ref = vcf[i, 4]
    cur.alt = vcf[i, 5]
    
    gene.idx = which(gene.info[, 2] <= cur.pos & gene.info[, 3] >= cur.pos)
    if(length(gene.idx) == 0)
      next
    
    eff.idx = pmatch('EFF=', cur.info)
    cur.eff = strsplit(cur.info[eff.idx], ',', fixed = T)[[1]]
    
    for(gi in gene.idx){
      gene = gene.info[gi, 1]
      
      cur.fc = '.'
      for(j in 1:length(cur.eff)){
        cur.eff.split = strsplit(cur.eff[j], '|', fixed = T)[[1]]
        # if(cur.eff.split[6] == gene){
        if(grepl(gene, cur.eff.split[6])){
          cur.fc = cur.eff.split[2]
          if(cur.fc == '')
            cur.fc = '.'
          break
        }
      }
      
      col.idx = which((out[[gene]][[1]][1,] == cur.pos) & (out[[gene]][[1]][2,] == cur.ref))
      if(length(col.idx) == 0){#new variant
        cur.col.genotype = c(cur.pos, cur.ref, rep(NA, n.row - 2))
        cur.col.genotype[row.idx] = cur.alt
        out[[gene]][[2]] = cbind(out[[gene]][[2]], cur.col.genotype)
        
        cur.col.fc = c(cur.pos, cur.ref, rep(NA, n.row - 2))
        cur.col.fc[row.idx] = cur.fc
        out[[gene]][[1]] = cbind(out[[gene]][[1]], cur.col.fc)
        
      }else{#variant exists
        out[[gene]][[2]][row.idx, col.idx] = cur.alt
        out[[gene]][[1]][row.idx, col.idx] = cur.fc
      }
    }
  }
  return(out)
}

extract.info <- function(work.dir = VCF_DIR, out.dir = EXTRACT_OUT_DIR, out.rdata.file = EXTRACT_OUT_RDATA_FILE, gene.info.file = GENE_INFO_FILE, vcf.file.post = VCF_FILE_POST, fasta.file = FASTA_FILE, to.exclude.samples.file = To_EXCLUDE_SAMPLES_FILE){
  # browser()
  
  system(paste('mkdir -p', work.dir))
  system(paste('mkdir -p', out.dir))
  setwd(work.dir)
  vcf.files = dir(pattern = '*.vcf$')
  
  grep.awk.cmd = paste0('grep ">" ', fasta.file, " | awk -F '>' '{print $2}' > fasta.samples.list")
  system(grep.awk.cmd)
  fasta.samples.list = read.table('fasta.samples.list', header = F, stringsAsFactors = F)[, 1]
  idx = vcf.files %in% paste0(fasta.samples.list, vcf.file.post)
  vcf.files = vcf.files[idx]
  system('rm fasta.samples.list')

  if(!is.null(to.exclude.samples.file)){
     to.exclude.samples = read.table(file = to.exclude.samples.file, header = F, stringsAsFactors = F)[, 1]
     idx = vcf.files %in% paste0(to.exclude.samples, vcf.file.post)
     vcf.files = vcf.files[!idx]
  }
  
  gene.info = read.table(file = gene.info.file, header = T, stringsAsFactors = F)
  genes = unique(gene.info[,1])
  out = vector('list', length(genes))
  names(out) = genes
  for(i in 1:length(out)) {
    out[[i]] = vector('list', 2)
    names(out[[i]]) = c('functional.class', 'genotype')
    for(j in 1:length(out[[i]]))
      out[[i]][[j]] = as.matrix(c('POS', 'REF', gsub(vcf.file.post, '', vcf.files)), ,1)
  }
  
  
  for(i in 1:length(vcf.files)){
    cat(i, '\t', vcf.files[i], '\n')
    cur.extraction = extract.functional.class.genotype(vcf.files[i], gene.info, out, vcf.file.post)
    if(is.null(cur.extraction))
      next
    out = cur.extraction
    cat('\n')
  }
  
  out.NA.removed = out
  
  for(i in 1:length(out)){
    cur.gene.genotype = out[[i]][[2]]
    flag = rep(TRUE, nrow(cur.gene.genotype))
    for(j in 1:nrow(cur.gene.genotype)){
      if(sum(!is.na(cur.gene.genotype[j, -1])) == 0)
        flag[j] = FALSE
    }
    out.NA.removed[[i]][[1]] = out.NA.removed[[i]][[1]][flag, ]
    out.NA.removed[[i]][[2]] = out.NA.removed[[i]][[2]][flag, ]
    
    colnames(out[[i]][[1]]) = NULL
    out[[i]][[1]] = data.frame(out[[i]][[1]], stringsAsFactors = F)
    
    colnames(out[[i]][[2]]) = NULL
    out[[i]][[2]] = data.frame(out[[i]][[2]], stringsAsFactors = F)
    
    colnames(out.NA.removed[[i]][[1]]) = NULL
    out.NA.removed[[i]][[1]] = data.frame(out.NA.removed[[i]][[1]], stringsAsFactors = F)
    
    colnames(out.NA.removed[[i]][[2]]) = NULL
    out.NA.removed[[i]][[2]] = data.frame(out.NA.removed[[i]][[2]], stringsAsFactors = F)
  }
  
  
  setwd(out.dir)
  save(out, file = out.rdata.file)
  # save(out, file = gsub('.Rdata', '.NA.removed.Rdata', out.rdata.file))
}
# extract.info()

significant.test <- function(pheno.vector, geno.matrix){
  # browser()
  
  pvalues.names = c('counts.glm', 'hits.glm', 'hits.fisher', 'skat', 'skat.optimal.adj')
  pvalues = rep(NA, length(pvalues.names))
  names(pvalues) = pvalues.names
  
  betas.names = c('counts.glm', 'hits.glm', 'hits.fisher', 'skat', 'skat.optimal.adj')
  betas = rep(NA, length(betas.names))
  names(betas) = betas.names
  
  counts = rowSums(geno.matrix)
  hits = as.numeric(counts > 0)
  
  res = tryCatch({
    summary(glm(pheno.vector ~ counts, family = binomial))$coefficients[2, c(1,4)]
  }, error = function(e) {
    return(c(NA, NA))
  })
  betas['counts.glm'] = res[1]
  pvalues['counts.glm'] = res[2]
  
  res = tryCatch({
    summary(glm(pheno.vector ~ hits, family = binomial))$coefficients[2, c(1, 4)]
  }, error = function(e) {
    return(c(NA, NA))
  })
  betas['hits.glm'] = res[1]
  pvalues['hits.glm'] = res[2]
  
  res = tryCatch({
    res.f = fisher.test(x = hits, y = pheno.vector)
	c(res.f$estimate, res.f$p.value)
  }, error = function(e) {
    return(c(NA, NA))
  })
  betas['hits.fisher'] = res[1]
  pvalues['hits.fisher'] = res[2]
  
  idx = colSums(geno.matrix) > 0
  if(sum(idx) > 0){
    library(SKAT)
    obj = SKAT_Null_Model(pheno.vector ~ 1, out_type="D", Adjustment = F)
    Z = as.matrix(geno.matrix[, idx, drop = F])
    
    res = tryCatch({
      res.s = SKAT(Z, obj, r.corr = 0)
      c(NA, res.s$p.value)
    }, error = function(e) {
      return(NA, NA)
    })
	betas['skat'] = res[1]
    pvalues['skat'] = res[2]
    
    res = tryCatch({
      res.so = SKAT(Z, obj, method = "optimal.adj")
	  c(NA, res.so$p.value)
    }, error = function(e) {
      return(NA, NA)
    })
	betas['skat.optimal.adj'] = res[1]
	pvalues['skat.optimal.adj'] = res[2]
  }
  significant.test.results = list(betas, pvalues)
  names(significant.test.results) = c('betas', 'pvalues')
  return(significant.test.results)
}

check.missing.alleles.scan <- function(cur.gene.genotype, pos.ref, fasta.file = FASTA_FILE){
  # browser()
  
  cur.gene.genotype = as.matrix(cur.gene.genotype)
  sample.names = rownames(cur.gene.genotype)
  for(i in 1:nrow(cur.gene.genotype)){
    cat('\n', nrow(cur.gene.genotype), '\t', i, '\n')
    
    cur.sample = sample.names[i]
    grep.cmd = paste0('grep -n ">', cur.sample, '" ', fasta.file, ' | cut -d ":" -f 1')
    cur.sample.line.idx = as.numeric(system(grep.cmd, intern = T)) + 1
    awk.cmd = paste0("awk '{if(NR==", cur.sample.line.idx, ") print}' ", fasta.file)
    cur.sample.fasta = system(awk.cmd, intern = T)
    
    for(j in 1:ncol(cur.gene.genotype)){
      cat(j, '\t')
      cur.pos = pos.ref[j, 1]
      real.allele = substr(cur.sample.fasta, cur.pos, cur.pos)
      
      if(is.na(cur.gene.genotype[i, j])){
        if(real.allele %in% c('A', 'T', 'G', 'C'))
          cur.gene.genotype[i, j] = real.allele
      }# else{
      # 				if(cur.gene.genotype[i, j] != real.allele)
      # 					warning(paste('allele conflicts:', i, cur.sample, j, cur.pos, sep = '\t'))
      # 			}
    }
  }
  # warnings()
  return(cur.gene.genotype)
}

viral.test <- function(cur.gene = 'E7', min.maf.thr = 0, max.maf.thr = 0.01, patterns = NULL, work.dir = EXTRACT_OUT_DIR, fc.g.rdata.file = EXTRACT_OUT_RDATA_FILE, pheno.file = PHENO_FILE, sig.genes = c('E7', 'E1', 'L1', 'E2', 'E4', 'E5', 'E6', 'L2', 'URR', 'E8'), lineages = c("AB","C"), status.name.in.pheno.file = STATUS_NAME_IN_PHENO_FILE, lineages.name.in.pheno.file = LINEAGES_NAME_IN_PHENO_FILE, out.dir = BURDEN_RESULTS_DIR){
  # browser()
  if(is.null(patterns))
    stop("No pattern input")
  #load pheno matrix, and extract the samples
  setwd(work.dir)
  pheno.info = read.delim(file = pheno.file, header = T, stringsAsFactors = F, sep = '\t', check.names = F)
  if(is.null(lineages.name.in.pheno.file))
    idx = (!is.na(pheno.info[, status.name.in.pheno.file]))
  else
    idx = (pheno.info[, lineages.name.in.pheno.file] %in% lineages) & (!is.na(pheno.info[, status.name.in.pheno.file]))
  
  pheno.info = pheno.info[idx,]
  
  load(fc.g.rdata.file)
  #first colum is POS, REF, sample1.vcf, sample2.vcf, ....
  #2nd colum to nth colum is the variant
  pos.ref = data.frame(t(out[[cur.gene]][['genotype']][1:2, -1]), stringsAsFactors = F)
  colnames(pos.ref) = c('POS', 'REF')
  # row.names = matrix(unlist(strsplit(out[[cur.gene]][['genotype']][-(1:2), 1], '_')), 2, )[1, ]
  row.names = out[[cur.gene]][['genotype']][-(1:2), 1]
  col.names = paste(pos.ref[, 'POS'], pos.ref[, 'REF'], sep = '_')
  
  idx = row.names %in% pheno.info[, 1]
  
  setwd(out.dir)
  
  cur.gene.functional.class = as.matrix(out[[cur.gene]][['functional.class']])[-(1:2), -1]
  rownames(cur.gene.functional.class) = row.names
  colnames(cur.gene.functional.class) = col.names
  cur.gene.functional.class = cur.gene.functional.class[idx, ]
  functional.class.Rdata.file = paste0(cur.gene, '_', paste(lineages, collapse = '_'), '.functional.class.Rdata')
  if(!file.exists(functional.class.Rdata.file))
    save(cur.gene.functional.class, file = functional.class.Rdata.file)
  
  cur.gene.genotype = as.matrix(out[[cur.gene]][['genotype']])[-(1:2), -1]
  rownames(cur.gene.genotype) = row.names
  colnames(cur.gene.genotype) = col.names
  cur.gene.genotype = cur.gene.genotype[idx, ]
  # browser()
  #mapping the IDs, so pheno and geno and variants matrix are in the same order
  idx = match(rownames(cur.gene.genotype), pheno.info[, 1])
  pheno.info = pheno.info[idx, ]
  
  #covert cell with pattern to 1, others to 0
  cur.gene.functional.class.hits = cur.gene.functional.class
  cur.gene.functional.class.hits[is.na(cur.gene.functional.class.hits)] = 0	
  idx = cur.gene.functional.class.hits %in% patterns
  cur.gene.functional.class.hits[idx] = 1
  cur.gene.functional.class.hits[!idx] = 0
  storage.mode(cur.gene.functional.class.hits) = 'integer'
  
  # checking missing alleles
  genotype.Rdata.file = paste0(cur.gene, '_', paste(lineages, collapse = '_'), '.genotype.Rdata')
  if(file.exists(genotype.Rdata.file))
    load(genotype.Rdata.file)
  else{
    cur.gene.genotype = check.missing.alleles.scan(cur.gene.genotype, pos.ref)
    save(cur.gene.genotype, file = genotype.Rdata.file)
  }
  
  
  cur.gene.genotype.hits = cur.gene.genotype
  af = matrix(NA, 5, ncol(cur.gene.genotype.hits))
  rownames(af) = c('REF_AlleleFreq', 'A_AlleleFreq', 'T_AlleleFreq', 'G_AlleleFreq', 'C_AlleleFreq')
  cur.gene.pos.ref.genotype.af = rbind(t(pos.ref), af, cur.gene.genotype.hits)
  
  
  alleles = c('A', 'T', 'G', 'C')
  cur.gene.genotype.hits[which(!cur.gene.genotype.hits %in% alleles)] = 0
  for(i in 1:ncol(cur.gene.genotype.hits)){
    n.alleles = length(which(cur.gene.genotype.hits[, i] %in% alleles))
    for(allele in alleles){
      idx = which(cur.gene.genotype.hits[, i] %in% allele)
      if(length(idx) < 1)
        next
      allele.freq = length(idx)/n.alleles
      
      if(allele.freq <= max.maf.thr & allele.freq > min.maf.thr)
        cur.gene.genotype.hits[idx, i] = 1
      else
        cur.gene.genotype.hits[idx, i] = 0
      
      if(allele == cur.gene.pos.ref.genotype.af[2, i])
        cur.gene.pos.ref.genotype.af[3, i] = allele.freq
      cur.gene.pos.ref.genotype.af[paste0(allele, '_AlleleFreq'), i] = allele.freq
    }
  }
  write.table(cur.gene.pos.ref.genotype.af, file = paste0(cur.gene, '_', paste(lineages, collapse = '_'), '.pos.ref.genotype.af.txt'), row.names = T, col.names = F, sep = '\t', quote = F)
  cat('cur.gene.pos.ref.genotype.af done!\n')
  
  storage.mode(cur.gene.genotype.hits) = 'integer'
  if(cur.gene == 'URR')
    final.hits = cur.gene.genotype.hits
  else
    final.hits = cur.gene.functional.class.hits * cur.gene.genotype.hits
  # sum.hits
  write.table(rowSums(final.hits), file = paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.hits.txt'), quote = F, sep = '\t', col.names = F, row.names = T)
  
  
  # browser()
  # variants distribution.summary
  if((!is.null(sig.genes)) & (cur.gene %in% sig.genes)){
    idx = which(colSums(final.hits) > 0)
    if(length(idx) < 1){
      variant.distribution.summary = matrix(0, 1, 5)
      colnames(variant.distribution.summary) = c('POS', 'REF', 'Functional.Class', 'Num.in.Controls', 'Num.in.Cases')			
    }else{
      variant.distribution.summary = matrix(0, length(idx), 5)
      colnames(variant.distribution.summary) = c('POS', 'REF', 'Functional.Class', 'Num.in.Controls', 'Num.in.Cases')
      variant.distribution.summary[, 1:2] = as.matrix(pos.ref[idx, ])
      for(i in 1:length(idx)){
        na.idx = is.na(cur.gene.functional.class[, idx[i]])
        variant.distribution.summary[i, 3] = paste(unique(cur.gene.functional.class[!na.idx, idx[i]]), collapse = '_')
        variant.distribution.summary[i, 4] = sum(final.hits[pheno.info[, status.name.in.pheno.file] == 0, idx[i]])
        variant.distribution.summary[i, 5] = sum(final.hits[pheno.info[, status.name.in.pheno.file] == 1, idx[i]])
      }
      
      idx = order(as.numeric(variant.distribution.summary[, 1]))
      variant.distribution.summary = variant.distribution.summary[idx, , drop=FALSE]
    }
    
    write.table(variant.distribution.summary, paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.variant.distribution.summary.txt'), sep = '\t', quote = F, col.names = T, row.names = F)
    # return('variants distribution.summary done!')
  }
  
  cat('variant.distribution.summary done!\n')
  
  #variants.summary
  CC.variants = table(pheno.info[, status.name.in.pheno.file], rowSums(final.hits))
  write.table(CC.variants, file = paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.variants.summary.txt'), quote = F, sep = '\t', col.names=NA)
  
  CC.variants = table(pheno.info[, status.name.in.pheno.file], as.numeric(rowSums(final.hits) > 0))
  write.table(CC.variants, file = paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.variants.hits.summary.txt'), quote = F, sep = '\t', col.names=NA)
  
  cat('variants.hits.summary done!\n')
  
  significant.test.results = significant.test(pheno.info[, status.name.in.pheno.file], final.hits)
  pvalues = significant.test.results$pvalues
  final.hits.pheno = cbind(final.hits, pheno.info[, status.name.in.pheno.file])
  colnames(final.hits.pheno) = paste0(cur.gene, '_', colnames(final.hits.pheno))
  colnames(final.hits.pheno)[ncol(final.hits.pheno)] = paste0(cur.gene, '_pheno')
  save(final.hits.pheno, file = paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.Rdata'))
  return(significant.test.results)
}

run.all.genes.test <- function(gene.info.file = GENE_INFO_FILE, work.dir = EXTRACT_OUT_DIR, out.dir = BURDEN_RESULTS_DIR, out.file.post = '.pvalue.txt', min.maf.thr = MIN_MAF_THR, max.maf.thr = MAX_MAF_THR, patterns = PATTERNS, lineages = LINEAGES, sig.genes = c('E7', 'E1', 'L1', 'E2', 'E4', 'E5', 'E6', 'L2', 'URR', 'E8')){
  
  setwd(work.dir)
  if(!file.exists(out.dir))
    system(paste('mkdir -p', out.dir))
  
  gene.info = read.table(file = gene.info.file, header = T, stringsAsFactors = F)
  genes = unique(gene.info[,1])
  pvalues = NULL
  betas = NULL
  # browser()
  for(i in 1:length(genes)){
    cur.gene = genes[i]
    cat(i, '\t', cur.gene, '\n')
    cur.significant.test.results = viral.test(cur.gene = cur.gene, max.maf.thr = max.maf.thr, min.maf.thr = min.maf.thr, patterns = patterns, lineages = lineages, sig.genes = sig.genes, out.dir = out.dir)
    # for all women
    # cur.pvalues = viral.test(cur.gene = cur.gene, max.maf.thr = max.maf.thr, min.maf.thr = min.maf.thr, patterns = patterns, lineages = lineages, sig.genes = sig.genes, out.dir = out.dir, lineages.name.in.pheno.file = NULL)
    pvalues = cbind(pvalues, cur.significant.test.results$pvalues)
    betas = cbind(betas, cur.significant.test.results$betas)
    tmp.hits.sum = read.table(file = paste0(cur.gene, '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.hits.txt'), header = F, stringsAsFactors = F)
    if(i == 1)
      hits.sum = tmp.hits.sum
    else{
      idx = match(hits.sum[, 1], tmp.hits.sum[,1])
      hits.sum = cbind(hits.sum, tmp.hits.sum[idx, 2])
    }	
    cat('\n')
  }
  setwd(out.dir)
  
  colnames(hits.sum) = c('sample', genes)
  write.table(hits.sum, file = paste0('All_Genes', '_min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), '.hits.txt'), sep = '\t', quote = F, col.names = T, row.names = F)
  
  pvalues = data.frame(genes, t(pvalues))
  colnames(pvalues) = c('Gene', names(cur.pvalues))
  write.table(pvalues, file = paste0('min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), out.file.post), col.names = T, row.names = F, quote = F, sep = '\t')
  
  betas = data.frame(genes, t(betas))
  colnames(betas) = c('Gene', names(cur.significant.test.results$betas))
  out.file.post = gsub('pvalue', 'beta', out.file.post)
  write.table(betas, file = paste0('min.maf.', min.maf.thr, '_', 'max.maf.', max.maf.thr, '_', paste(patterns, collapse = '_'), '_', paste(lineages, collapse = '_'), out.file.post), col.names = T, row.names = F, quote = F, sep = '\t')
}
# run.all.genes.test(min.maf.thr = 0, max.maf.thr = 0.03, patterns = c("MISSENSE", "NONSENSE"), lineages = c("AB","C"))


combine.genes.analysis <- function(work.dir = BURDEN_RESULTS_DIR, final.hits.pheno.Rdata.file.post = FINAL_HITS_PHENO_RDATA_FILE_POST, genes.list = c('E7', 'E1', 'L1', 'E2', 'E4', 'E5', 'E6', 'L2', 'URR', 'E8')){
	# browser()
	setwd(work.dir)
	final.hits = NULL
	pheno = NULL
	
	cat("Gene", 'Num.Samples', 'Num.Variants.in.Controls', 'Num.Controls.with.Variants', 'Num.Variants.in.Cases', 'Num.Cases.with.Variants\n',sep = '\t')
	
	for(i in 1:length(genes.list)){
		final.hits.pheno.Rdata.file = paste0(genes.list[i], final.hits.pheno.Rdata.file.post)
		load(final.hits.pheno.Rdata.file)
		cat(genes.list[i], '\t', nrow(final.hits.pheno), '\t')
		
		if(is.null(final.hits)){
			final.hits = final.hits.pheno[, -ncol(final.hits.pheno)]
			pheno = final.hits.pheno[, ncol(final.hits.pheno)]
		}else{
			idx = match(rownames(final.hits), rownames(final.hits.pheno))
			final.hits = cbind(final.hits, final.hits.pheno[idx, -ncol(final.hits.pheno)])
			pheno = final.hits.pheno[idx, ncol(final.hits.pheno)]
		}
		
		cur.gene.final.hits = final.hits[, grepl(paste0(genes.list[i], "_"), colnames(final.hits))]
		
		table.cur.gene.final.hits = table(pheno, rowSums(cur.gene.final.hits))
		Num.Variants.in.Controls = sum(table.cur.gene.final.hits[1,] * as.numeric(colnames(table.cur.gene.final.hits)))
		Num.Variants.in.Cases = sum(table.cur.gene.final.hits[2,] * as.numeric(colnames(table.cur.gene.final.hits)))
		Num.Controls.with.Variants = sum(table.cur.gene.final.hits[1, -1])
		Num.Cases.with.Variants = sum(table.cur.gene.final.hits[2, -1])
		cat(Num.Variants.in.Controls, Num.Controls.with.Variants, Num.Variants.in.Cases, Num.Cases.with.Variants, '\n', sep = '\t')
	}
	
	cat('\n\nAll Genes\n\n')
	table.pheno.final.hits = table(pheno, rowSums(final.hits))
	Num.Variants.in.Controls = sum(table.pheno.final.hits[1,] * as.numeric(colnames(table.pheno.final.hits)))
	Num.Variants.in.Cases = sum(table.pheno.final.hits[2,] * as.numeric(colnames(table.pheno.final.hits)))
	Num.Controls.with.Variants = sum(table.pheno.final.hits[1, -1])
	Num.Cases.with.Variants = sum(table.pheno.final.hits[2, -1])
	cat('Num.Variants.in.Controls', 'Num.Controls.with.Variants', 'Num.Variants.in.Cases', 'Num.Cases.with.Variants\n',sep = '\t')
	cat(Num.Variants.in.Controls, Num.Controls.with.Variants, Num.Variants.in.Cases, Num.Cases.with.Variants, '\n', sep = '\t')
	
	significant.test.results = significant.test(pheno, final.hits)
	
	cat('\n\nAll Genes Combind Test Pvalues:\n\n')
	pvalues = significant.test.results$pvalues
	cat(names(pvalues), '\n', sep = '\t')
	cat(pvalues,'\n', sep = '\t')
	# browser()

	cat('\n\nAll Genes Combind Test Betas:\n\n')
	betas = significant.test.results$betas
	cat(names(betas), '\n', sep = '\t')
	cat(betas,'\n', sep = '\t')
}

extract.info()
run.all.genes.test()
combine.genes.analysis()
